pkg/ingester/label_names_and_values.go

```diff
@@ -136,6 +136,25 @@ func labelValuesCardinality(
 		if err != nil {
 			return err
 		}
+
+		// Check if empty value for label is allowed according to matchers. This
+		// is an optimization only to avoid fetching postings, since the posting
+		// list would be empty anyway.
+		emptyValueAllowedByMatchers := true
+		for _, matcher := range matchers {
+			if matcher.Name != lblName {
+				continue
+			}
+			if !matcher.Matches("") {
+				emptyValueAllowedByMatchers = false
+				break
+			}
+		}
+
+		if emptyValueAllowedByMatchers {
+			lblValues = append(lblValues, "")
+		}
```

Comment:
nitpick if you can move this into a tiny function of its own because `labelValuesCardinality` is becoming long with time